docs/man/Server.3
.TH "Server" 3 "30 Apr 2004" "InspIRCd" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Server \- Allows server output and query functions This class contains methods which allow a module to query the state of the irc server, and produce output to users and other servers. 

.PP
.SH SYNOPSIS
.br
.PP
\fC#include <modules.h>\fP
.PP
Inherits \fBclassbase\fP.
.PP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBServer\fP ()"
.br
.RI "\fIDefault constructor.\fP"
.ti -1c
.RI "virtual \fB~Server\fP ()"
.br
.RI "\fIDefault destructor.\fP"
.ti -1c
.RI "virtual void \fBSendOpers\fP (std::string s)"
.br
.RI "\fISends text to all opers.\fP"
.ti -1c
.RI "virtual void \fBLog\fP (int level, std::string s)"
.br
.RI "\fIWrites a log string.\fP"
.ti -1c
.RI "virtual void \fBSend\fP (int Socket, std::string s)"
.br
.RI "\fISends a line of text down a TCP/IP socket.\fP"
.ti -1c
.RI "virtual void \fBSendServ\fP (int Socket, std::string s)"
.br
.RI "\fISends text from the server to a socket.\fP"
.ti -1c
.RI "virtual void \fBSendFrom\fP (int Socket, \fBuserrec\fP *User, std::string s)"
.br
.RI "\fISends text from a user to a socket.\fP"
.ti -1c
.RI "virtual void \fBSendTo\fP (\fBuserrec\fP *Source, \fBuserrec\fP *Dest, std::string s)"
.br
.RI "\fISends text from a user to another user.\fP"
.ti -1c
.RI "virtual void \fBSendChannel\fP (\fBuserrec\fP *User, \fBchanrec\fP *Channel, std::string s, bool IncludeSender)"
.br
.RI "\fISends text from a user to a channel (mulicast).\fP"
.ti -1c
.RI "virtual bool \fBCommonChannels\fP (\fBuserrec\fP *u1, \fBuserrec\fP *u2)"
.br
.RI "\fIReturns true if two users share a common channel.\fP"
.ti -1c
.RI "virtual void \fBSendCommon\fP (\fBuserrec\fP *User, std::string text, bool IncludeSender)"
.br
.RI "\fISends text from a user to one or more channels (mulicast).\fP"
.ti -1c
.RI "virtual void \fBSendWallops\fP (\fBuserrec\fP *User, std::string text)"
.br
.RI "\fISends a WALLOPS message.\fP"
.ti -1c
.RI "virtual bool \fBIsNick\fP (std::string nick)"
.br
.RI "\fIReturns true if a nick is valid.\fP"
.ti -1c
.RI "virtual int \fBCountUsers\fP (\fBchanrec\fP *c)"
.br
.RI "\fIReturns a count of the number of users on a channel.\fP"
.ti -1c
.RI "virtual \fBuserrec\fP * \fBFindNick\fP (std::string nick)"
.br
.RI "\fIAttempts to look up a nick and return a pointer to it.\fP"
.ti -1c
.RI "virtual \fBchanrec\fP * \fBFindChannel\fP (std::string channel)"
.br
.RI "\fIAttempts to look up a channel and return a pointer to it.\fP"
.ti -1c
.RI "virtual std::string \fBChanMode\fP (\fBuserrec\fP *User, \fBchanrec\fP *Chan)"
.br
.RI "\fIAttempts to look up a user's privilages on a channel.\fP"
.ti -1c
.RI "virtual std::string \fBGetServerName\fP ()"
.br
.RI "\fIReturns the server name of the server where the module is loaded.\fP"
.ti -1c
.RI "virtual std::string \fBGetNetworkName\fP ()"
.br
.RI "\fIReturns the network name, global to all linked servers.\fP"
.ti -1c
.RI "virtual \fBAdmin\fP \fBGetAdmin\fP ()"
.br
.RI "\fIReturns the information of the server as returned by the /ADMIN command.\fP"
.ti -1c
.RI "virtual bool \fBAddExtendedMode\fP (char modechar, int type, bool requires_oper, int params_when_on, int params_when_off)"
.br
.RI "\fIAdds an extended mode letter which is parsed by a module This allows modules to add extra mode letters, e.g.\fP"
.ti -1c
.RI "virtual void \fBAddCommand\fP (char *cmd, handlerfunc f, char flags, int minparams)"
.br
.RI "\fIAdds a command to the command table.\fP"
.ti -1c
.RI "virtual void \fBSendMode\fP (char **parameters, int pcnt, \fBuserrec\fP *user)"
.br
.RI "\fISends a servermode.\fP"
.ti -1c
.RI "virtual void \fBSendToModeMask\fP (std::string modes, int flags, std::string text)"
.br
.RI "\fISends to all users matching a mode mask You must specify one or more usermodes as the first parameter.\fP"
.ti -1c
.RI "virtual \fBchanrec\fP * \fBJoinUserToChannel\fP (\fBuserrec\fP *user, std::string cname, std::string key)"
.br
.RI "\fIForces a user to join a channel.\fP"
.ti -1c
.RI "virtual \fBchanrec\fP * \fBPartUserFromChannel\fP (\fBuserrec\fP *user, std::string cname, std::string reason)"
.br
.RI "\fIForces a user to part a channel.\fP"
.ti -1c
.RI "virtual void \fBChangeUserNick\fP (\fBuserrec\fP *user, std::string nickname)"
.br
.RI "\fIForces a user nickchange.\fP"
.ti -1c
.RI "virtual void \fBQuitUser\fP (\fBuserrec\fP *user, std::string reason)"
.br
.RI "\fIForces a user to quit with the specified reason.\fP"
.ti -1c
.RI "virtual bool \fBMatchText\fP (std::string sliteral, std::string spattern)"
.br
.RI "\fIMatches text against a glob pattern.\fP"
.ti -1c
.RI "virtual void \fBCallCommandHandler\fP (std::string commandname, char **parameters, int pcnt, \fBuserrec\fP *user)"
.br
.RI "\fICalls the handler for a command, either implemented by the core or by another module.\fP"
.ti -1c
.RI "virtual void \fBChangeHost\fP (\fBuserrec\fP *user, std::string host)"
.br
.RI "\fIChange displayed hostname of a user.\fP"
.ti -1c
.RI "virtual void \fBChangeGECOS\fP (\fBuserrec\fP *user, std::string gecos)"
.br
.RI "\fIChange GECOS (fullname) of a user.\fP"
.ti -1c
.RI "virtual bool \fBIsUlined\fP (std::string server)"
.br
.RI "\fIReturns true if the servername you give is ulined.\fP"
.in -1c
.SH "Detailed Description"
.PP 
Allows server output and query functions This class contains methods which allow a module to query the state of the irc server, and produce output to users and other servers.

All modules should instantiate at least one copy of this class, and use its member functions to perform their tasks. 
.PP
Definition at line 252 of file modules.h.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Server::Server ()"
.PP
Default constructor.Creates a Server object.Definition at line 135 of file modules.cpp.
.PP
.nf
136 {
137 }
.fi
.SS "Server::~Server ()\fC [virtual]\fP"
.PP
Default destructor.Destroys a Server object.Definition at line 139 of file modules.cpp.
.PP
.nf
140 {
141 }
.fi
.SH "Member Function Documentation"
.PP 
.SS "void Server::AddCommand (char * cmd, handlerfunc f, char flags, int minparams)\fC [virtual]\fP"
.PP
Adds a command to the command table.This allows modules to add extra commands into the command table. You must place a function within your module which is is of type handlerfunc:
.PP
typedef void (handlerfunc) (char**, int, userrec*); ... void \fBhandle_kill(char **parameters, int pcnt, userrec *user)\fP
.PP
When the command is typed, the parameters will be placed into the parameters array (similar to argv) and the parameter count will be placed into pcnt (similar to argv). There will never be any less parameters than the 'minparams' value you specified when creating the command. The *user parameter is the class of the user which caused the command to trigger, who will always have the flag you specified in 'flags' when creating the initial command. For example to create an oper only command create the commands with flags='o'.Definition at line 197 of file modules.cpp.
.PP
References createcommand().
.PP
.nf
198 {
199         createcommand(cmd,f,flags,minparams);
200 }
.fi
.SS "bool Server::AddExtendedMode (char modechar, int type, bool requires_oper, int params_when_on, int params_when_off)\fC [virtual]\fP"
.PP
Adds an extended mode letter which is parsed by a module This allows modules to add extra mode letters, e.g.+x for hostcloak. the 'type' parameter is either MT_CHANNEL, MT_CLIENT, or MT_SERVER, to indicate wether the mode is a channel mode, a client mode, or a server mode. requires_oper is used with MT_CLIENT type modes only to indicate the mode can only be set or unset by an oper. If this is used for MT_CHANNEL type modes it is ignored. params_when_on is the number of modes to expect when the mode is turned on (for type MT_CHANNEL only), e.g. with mode +k, this would have a value of 1. the params_when_off value has a similar value to params_when_on, except it indicates the number of parameters to expect when the mode is disabled. Modes which act in a similar way to channel mode +l (e.g. require a parameter to enable, but not to disable) should use this parameter. The function returns false if the mode is unavailable, and will not attempt to allocate another character, as this will confuse users. This also means that as only one module can claim a specific mode character, the core does not need to keep track of which modules own which modes, which speeds up operation of the server. In this version, a mode can have at most one parameter, attempting to use more parameters will have undefined effects.Definition at line 317 of file modules.cpp.
.PP
References DoAddExtendedMode(), MT_CLIENT, and MT_SERVER.
.PP
.nf
318 {
319         if (type == MT_SERVER)
320         {
321                 log(DEBUG,'*** API ERROR *** Modes of type MT_SERVER are reserved for future expansion');
322                 return false;
323         }
324         if (((params_when_on>0) || (params_when_off>0)) && (type == MT_CLIENT))
325         {
326                 log(DEBUG,'*** API ERROR *** Parameters on MT_CLIENT modes are not supported');
327                 return false;
328         }
329         if ((params_when_on>1) || (params_when_off>1))
330         {
331                 log(DEBUG,'*** API ERROR *** More than one parameter for an MT_CHANNEL mode is not yet supported');
332                 return false;
333         }
334         return DoAddExtendedMode(modechar,type,requires_oper,params_when_on,params_when_off);
335 }
.fi
.SS "void Server::CallCommandHandler (std::string commandname, char ** parameters, int pcnt, \fBuserrec\fP * user)\fC [virtual]\fP"
.PP
Calls the handler for a command, either implemented by the core or by another module.You can use this function to trigger other commands in the ircd, such as PRIVMSG, JOIN, KICK etc, or even as a method of callback. By defining command names that are untypeable for users on irc (e.g. those which contain a  or 
.br
) you may use them as callback identifiers. The first parameter to this method is the name of the command handler you wish to call, e.g. PRIVMSG. This will be a command handler previously registered by the core or wih \fBAddCommand()\fP. The second parameter is an array of parameters, and the third parameter is a count of parameters in the array. If you do not pass enough parameters to meet the minimum needed by the handler, the functiom will silently ignore it. The final parameter is the user executing the command handler, used for privilage checks, etc.Definition at line 187 of file modules.cpp.
.PP
.nf
188 {
189         call_handler(commandname.c_str(),parameters,pcnt,user);
190 }
.fi
.SS "void Server::ChangeGECOS (\fBuserrec\fP * user, std::string gecos)\fC [virtual]\fP"
.PP
Change GECOS (fullname) of a user.You should always call this method to change a user's GECOS rather than writing directly to the fullname member of userrec, as any change applied via this method will be propogated to any linked servers.Definition at line 275 of file modules.cpp.
.PP
References ChangeName().
.PP
.nf
276 {
277         ChangeName(user,gecos.c_str());
278 }
.fi
.SS "void Server::ChangeHost (\fBuserrec\fP * user, std::string host)\fC [virtual]\fP"
.PP
Change displayed hostname of a user.You should always call this method to change a user's host rather than writing directly to the dhost member of userrec, as any change applied via this method will be propogated to any linked servers.Definition at line 270 of file modules.cpp.
.PP
References ChangeDisplayedHost().
.PP
.nf
271 {
272         ChangeDisplayedHost(user,host.c_str());
273 }
.fi
.SS "void Server::ChangeUserNick (\fBuserrec\fP * user, std::string nickname)\fC [virtual]\fP"
.PP
Forces a user nickchange.This command works similarly to SVSNICK, and can be used to implement Q-lines etc. If you specify an invalid nickname, the nick change will be dropped and the target user will receive the error numeric for it.Definition at line 171 of file modules.cpp.
.PP
.nf
172 {
173         force_nickchange(user,nickname.c_str());
174 }
.fi
.SS "std::string Server::ChanMode (\fBuserrec\fP * User, \fBchanrec\fP * Chan)\fC [virtual]\fP"
.PP
Attempts to look up a user's privilages on a channel.This function will return a string containing either @, , +, or an empty string, representing the user's privilages upon the channel you specify.Definition at line 295 of file modules.cpp.
.PP
References cmode().
.PP
.nf
296 {
297         return cmode(User,Chan);
298 }
.fi
.SS "bool Server::CommonChannels (\fBuserrec\fP * u1, \fBuserrec\fP * u2)\fC [virtual]\fP"
.PP
Returns true if two users share a common channel.This method is used internally by the NICK and QUIT commands, and the \fBServer::SendCommon\fP method.Definition at line 248 of file modules.cpp.
.PP
References common_channels().
.PP
.nf
249 {
250         return (common_channels(u1,u2) != 0);
251 }
.fi
.SS "int Server::CountUsers (\fBchanrec\fP * c)\fC [virtual]\fP"
.PP
Returns a count of the number of users on a channel.This will NEVER be 0, as if the chanrec exists, it will have at least one user in the channel.Definition at line 337 of file modules.cpp.
.PP
.nf
338 {
339         return usercount(c);
340 }
.fi
.SS "\fBchanrec\fP * Server::FindChannel (std::string channel)\fC [virtual]\fP"
.PP
Attempts to look up a channel and return a pointer to it.This function will return NULL if the channel does not exist.Definition at line 290 of file modules.cpp.
.PP
.nf
291 {
292         return FindChan(channel.c_str());
293 }
.fi
.SS "\fBuserrec\fP * Server::FindNick (std::string nick)\fC [virtual]\fP"
.PP
Attempts to look up a nick and return a pointer to it.This function will return NULL if the nick does not exist.Definition at line 285 of file modules.cpp.
.PP
.nf
286 {
287         return Find(nick);
288 }
.fi
.SS "\fBAdmin\fP Server::GetAdmin ()\fC [virtual]\fP"
.PP
Returns the information of the server as returned by the /ADMIN command.See the \fBAdmin\fP class for further information of the return value. The members \fBAdmin::Nick\fP, \fBAdmin::Email\fP and \fBAdmin::Name\fP contain the information for the server where the module is loaded.Definition at line 310 of file modules.cpp.
.PP
.nf
311 {
312         return Admin(getadminname(),getadminemail(),getadminnick());
313 }
.fi
.SS "std::string Server::GetNetworkName ()\fC [virtual]\fP"
.PP
Returns the network name, global to all linked servers.Definition at line 305 of file modules.cpp.
.PP
.nf
306 {
307         return getnetworkname();
308 }
.fi
.SS "std::string Server::GetServerName ()\fC [virtual]\fP"
.PP
Returns the server name of the server where the module is loaded.Definition at line 300 of file modules.cpp.
.PP
.nf
301 {
302         return getservername();
303 }
.fi
.SS "bool Server::IsNick (std::string nick)\fC [virtual]\fP"
.PP
Returns true if a nick is valid.Nicks for unregistered connections will return false.Definition at line 280 of file modules.cpp.
.PP
References isnick().
.PP
.nf
281 {
282         return (isnick(nick.c_str()) != 0);
283 }
.fi
.SS "bool Server::IsUlined (std::string server)\fC [virtual]\fP"
.PP
Returns true if the servername you give is ulined.ULined servers have extra privilages. They are allowed to change nicknames on remote servers, change modes of clients which are on remote servers and set modes of channels where there are no channel operators for that channel on the ulined server, amongst other things. Ulined server data is also broadcast across the mesh at all times as opposed to selectively messaged in the case of normal servers, as many ulined server types (such as services) do not support meshed links and must operate in this manner.Definition at line 182 of file modules.cpp.
.PP
References is_uline().
.PP
.nf
183 {
184         return is_uline(server.c_str());
185 }
.fi
.SS "\fBchanrec\fP * Server::JoinUserToChannel (\fBuserrec\fP * user, std::string cname, std::string key)\fC [virtual]\fP"
.PP
Forces a user to join a channel.This is similar to svsjoin and can be used to implement redirection, etc. On success, the return value is a valid pointer to a chanrec* of the channel the user was joined to. On failure, the result is NULL.Definition at line 161 of file modules.cpp.
.PP
.nf
162 {
163         return add_channel(user,cname.c_str(),key.c_str(),true);
164 }
.fi
.SS "void Server::Log (int level, std::string s)\fC [virtual]\fP"
.PP
Writes a log string.This method writes a line of text to the log. If the level given is lower than the level given in the configuration, this command has no effect.Definition at line 192 of file modules.cpp.
.PP
.nf
193 {
194         log(level,'%s',s.c_str());
195 }
.fi
.SS "bool Server::MatchText (std::string sliteral, std::string spattern)\fC [virtual]\fP"
.PP
Matches text against a glob pattern.Uses the ircd's internal matching function to match string against a globbing pattern, e.g. *!*@*.com Returns true if the literal successfully matches the pattern, false if otherwise.Definition at line 148 of file modules.cpp.
.PP
.nf
149 {
150         char literal[MAXBUF],pattern[MAXBUF];
151         strncpy(literal,sliteral.c_str(),MAXBUF);
152         strncpy(pattern,spattern.c_str(),MAXBUF);
153         return match(literal,pattern);
154 }
.fi
.SS "\fBchanrec\fP * Server::PartUserFromChannel (\fBuserrec\fP * user, std::string cname, std::string reason)\fC [virtual]\fP"
.PP
Forces a user to part a channel.This is similar to svspart and can be used to implement redirection, etc. Although the return value of this function is a pointer to a channel record, the returned data is undefined and should not be read or written to. This behaviour may be changed in a future version.Definition at line 166 of file modules.cpp.
.PP
.nf
167 {
168         return del_channel(user,cname.c_str(),reason.c_str(),false);
169 }
.fi
.SS "void Server::QuitUser (\fBuserrec\fP * user, std::string reason)\fC [virtual]\fP"
.PP
Forces a user to quit with the specified reason.To the user, it will appear as if they typed /QUIT themselves, except for the fact that this function may bypass the quit prefix specified in the config file.
.PP
WARNING!
.PP
Once you call this function, userrec* user will immediately become INVALID. You MUST NOT write to, or read from this pointer after calling the QuitUser method UNDER ANY CIRCUMSTANCES! The best course of action after calling this method is to immediately bail from your handler.Definition at line 176 of file modules.cpp.
.PP
References userrec::nick, and send_network_quit().
.PP
.nf
177 {
178         send_network_quit(user->nick,reason.c_str());
179         kill_link(user,reason.c_str());
180 }
.fi
.SS "void Server::Send (int Socket, std::string s)\fC [virtual]\fP"
.PP
Sends a line of text down a TCP/IP socket.This method writes a line of text to an established socket, cutting it to 510 characters plus a carriage return and linefeed if required.Definition at line 207 of file modules.cpp.
.PP
.nf
208 {
209         Write(Socket,'%s',s.c_str());
210 }
.fi
.SS "void Server::SendChannel (\fBuserrec\fP * User, \fBchanrec\fP * Channel, std::string s, bool IncludeSender)\fC [virtual]\fP"
.PP

docs/ChangeLog
1.0 Final

    * Optimized WriteServ - why use %s!%s@%s when we have the more optimal userrec::GetFullHost() ?
    * Added new event into m_spanningtree, event id send_metadata, which allows arbitary posting of metadata outside of the netburst, see line 529 of m_ssl_gnutls for an example of how to use it.
    * Added check that forbids configuration as root >:-)
    * Major updates to CBANs...like they're reasonably functional
    * Added support for TYPE_OTHER metadata, where void* target will be NULL, e.g. its not bound to a user or a nick
    * Fix to send 403 on invalid channel name
    * Moved Write() call so that it can send the NICK message properly and force the client to change
    * Allow ulines to still notice the channel regardless of +T
    * Created OnPostOper that can be gauranteed to be called *after* opering (by ALL modules) is done
    * "Oper of unknown type" -> "313 %s %s :is an IRC operator"
    * Auto-assume "+" on the start of opermode block if not included
    * Removed example commands that never actually existed
    * OpenSSL detection and cert generation in ./configure
    * Changed length of password field to 64, giving us a maximum password length of 63 plus null terminator
    * Fixed uninitialized ExtMode::list
    * Added some extra debug
    * Make m_services use InsertMode (are we done yet?)
    * Changed On005Numeric to use InsertMode (we left one out?)
    * Deleted the semicolon that had half the dev team wondering...
    * Incorporated Craig's friendlier SQL class (documentation to come later)
    * m_sqlauth: Fix small memory leaks in some cases (bad password, bad query, lost mysql server connection etc..)
    * Correctly initialize userrec::password (should fix a few m_sql issues)
    * Add m_conn_waitpong, require clients to reply with a correct PONG to a server PING before they can connect
    * Remove unused includes and rename classes from ModuleSSL to ModuleSSLGnuTLS so we can be consistent across multiple SSL modules
    * Added warnings to opers when session limits are exceeded
    * Improved IP handling. Now uses in_addr to store client ip, not char[16]!, Added global and local session limits
    * Problems removing +t due to wrong bit-value being used in comparison
    * Altered so that if running in -debug mode, will not erase its .so files from /tmp to allow a clean backtrace
    * Removed unneccessary strlcpy
    * Removed a load of unneeded includes (poll.h etc)
    * Added <ident:timeout> example so that people remember to set this when using m_ident.so
    * Default ident timeout to 1 sec if not defined (a bit low, but better than 0!)
    * Updated m_ident to use inspsocket, should be more reliable
    * Added missing modules (newer ones)
    * Removed excessive debug for something thats now fixed
    * Merged Andy Church's isnick() and isident() patches - faster, more efficient, and doesnt accidentally allow "-" as the first character of a nickname, Also adapted to iterate over a char* pointer rather than using unsigned int for the for() loops
    * Added m_opersha256 module created by Special
    * Add module for anti-autorejoin channel mode +J
    * Changes to INVEX and EXCEPTS 005 tokens
    * Changed +l to drop the mode if the limit is same as current., Optimized check on numeric chars
    * Remove last relic of last goto in inspircd (an unused label)
    * Fixed to not have ohv in CHANMODES type A,, Added proper MAXMODES output in 005 instead of hard coded '13'
    * One goto is now no gotos.
    * Change m_redirect, m_joinflood and m_messageflood to put their modes in the correct section and to use InsertMode()
    * Add InsertMode function to helperfuncs.(cpp|h) for easily adding modes to the CHANMODES section of the 005 numeric
    * Added patch from Andy Church re spaces in SETNAME
    * Fixed line endings when skipping ssl cert generation
    * Documented some functions that were missing comments
    * Properly check that existing key resides in conf/ and not current dir
    * Added PRIORITY_BEFORE and PRIORITY_AFTER (see src/modules/m_hostchange.cpp for how it works, function Prioritize())
    * Added -modupdate to regenerate modules makefile, Added SSL option, as well as a 'Build Certificate' section (If avaliable)
    * Adding the +e and +I modules m_banexception and m_inviteexception, also adding their shared utility header
    * Added trick to allow local server to unset -r only in one specific condition (the nickchange)
    * Added joinflood +j
    * Made bans check both real and fake host
    * Patched to allow FTOPIC from users (so that chanserv can update the settime and the topic still come from chanserv)
    * Added <link:hidden> to docs
    * Optional IP-Hiding for servers
    * Use libgnutls-config --libs to configure linker options
    * Make relative paths in the config file tags be relative to the insp config directory
    * Change 974 numeric to 490 to avoid collision with Insp's failed to load module error
    * Make m_sslmodes messages appear in the channel window
    * Updated with new docs for maxtargets
    * Updated 005 to show changeable MAXTARGETS
    * Allowed for configuration of maxtargets in <options:maxtargets>
    * Wrong nicklen is shown in the 005 numeric, fixed (noticed by ThaPrince)
    * Initial commit of m_sslmodes, supplying chanmode +z
    * Allow for <bind> tags with no or empty type, as well as 'clients' ones
    * Set vendor and static module flags...even though the booms from trying to unload it would be fun sad.gif
    * Make errors loading certificates DEFAULT level
    * Inital commit of the first SSL module \o/ \o/ \o/ Also Ommeh's first official svn commit \o/
    * Added MakeWildHost to userrec (*!*@dhost), Added ban stuff to messageflood
    * Fixes to server kick (i hope)
    * And yae, the compiler sayeth to brain, fix thine warning. and thou warning art fixeth.
    * Made '*' ban-specifier work

src/modules/m_spanningtree/server.cpp
/*
 * InspIRCd -- Internet Relay Chat Daemon
 *
 *   Copyright (C) 2017-2020, 2022-2025 Sadie Powell <sadie@sadiepowell.dev>
 *   Copyright (C) 2013-2016 Attila Molnar <attilamolnar@hush.com>
 *   Copyright (C) 2012, 2019 Robby <robby@chatbelgie.be>
 *   Copyright (C) 2009-2010 Daniel De Graaf <danieldg@inspircd.org>
 *   Copyright (C) 2009 Uli Schlachter <psychon@znc.in>
 *   Copyright (C) 2008-2010 Craig Edwards <brain@inspircd.org>
 *   Copyright (C) 2008 Robin Burchell <robin+git@viroteck.net>
 *
 * This file is part of InspIRCd.  InspIRCd is free software: you can
 * redistribute it and/or modify it under the terms of the GNU General Public
 * License as published by the Free Software Foundation, version 2.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


#include <filesystem>

#include "inspircd.h"
#include "modules/ssl.h"

#include "main.h"
#include "utils.h"
#include "link.h"
#include "treeserver.h"
#include "treesocket.h"
#include "commands.h"

namespace
{
	bool IsLocalRange(const irc::sockets::sockaddrs& sa)
	{
		if (sa.is_local())
			return true; // Always allowed.

		std::error_code ec;
		if (Utils->LocalRanges.empty() && std::filesystem::is_regular_file("/.dockerenv", ec))
			return true; // TODO: remove this a future release.

		for (const auto& cidr : Utils->LocalRanges)
		{
			if (cidr.match(sa))
				return true; // Explicitly allowed range.
		}
		return false;
	}
}

/*
 * Some server somewhere in the network introducing another server.
 *	-- w
 */
CmdResult CommandServer::HandleServer(TreeServer* ParentOfThis, Params& params)
{
	const std::string& servername = params[0];
	const std::string& sid = params[1];
	const std::string& description = params.back();
	TreeSocket* socket = ParentOfThis->GetSocket();

	if (!InspIRCd::IsSID(sid))
	{
		socket->SendError("Invalid format server ID: "+sid+"!");
		return CmdResult::FAILURE;
	}
	TreeServer* CheckDupe = Utils->FindServer(servername);
	if (CheckDupe)
	{
		socket->SendError("Server "+servername+" already exists!");
		ServerInstance->SNO.WriteToSnoMask('L', "Server \002"+CheckDupe->GetName()+"\002 being introduced from \002" + ParentOfThis->GetName() + "\002 denied, already exists. Closing link with " + ParentOfThis->GetName());
		return CmdResult::FAILURE;
	}
	CheckDupe = Utils->FindServer(sid);
	if (CheckDupe)
	{
		socket->SendError("Server ID "+sid+" already exists! You may want to specify the server ID for the server manually with <server:id> so they do not conflict.");
		ServerInstance->SNO.WriteToSnoMask('L', "Server \002"+servername+"\002 being introduced from \002" + ParentOfThis->GetName() + "\002 denied, server ID already exists on the network. Closing link with " + ParentOfThis->GetName());
		return CmdResult::FAILURE;
	}

	TreeServer* route = ParentOfThis->GetRoute();
	std::shared_ptr<Link> lnk = Utils->FindLink(route->GetName());
	auto* Node = new TreeServer(servername, description, sid, ParentOfThis, ParentOfThis->GetSocket(), lnk ? lnk->Hidden : false);

	HandleExtra(Node, params);

	ServerInstance->SNO.WriteToSnoMask('L', "Server \002"+ParentOfThis->GetName()+"\002 introduced server \002"+servername+"\002 ("+description+")");
	return CmdResult::SUCCESS;
}

void CommandServer::HandleExtra(TreeServer* newserver, Params& params)
{
	for (CommandBase::Params::const_iterator i = params.begin() + 2; i != params.end() - 1; ++i)
	{
		const std::string& prop = *i;
		std::string::size_type p = prop.find('=');

		std::string key = prop;
		std::string val;
		if (p != std::string::npos)
		{
			key.erase(p);
			val.assign(prop, p+1, std::string::npos);
		}

		if (irc::equals(key, "burst"))
			newserver->BeginBurst(ConvToNum<uint64_t>(val));
		else if (irc::equals(key, "hidden"))
			newserver->Hidden = !!ConvToNum<uint8_t>(val);
	}
}

std::shared_ptr<Link> TreeSocket::AuthRemote(const CommandBase::Params& params)
{
	size_t offset = proto_version == PROTO_INSPIRCD_3;
	if (params.size() < 4+offset)
	{
		SendError("Protocol error - Not enough parameters for SERVER command");
		return nullptr;
	}

	const std::string& sname = params[0];
	const std::string& password = params[1];
	const std::string& sid = params[2+offset];
	const std::string& description = params.back();

	this->SendCapabilities(2);

	if (!InspIRCd::IsSID(sid))
	{
		this->SendError("Invalid format server ID: "+sid+"!");
		return nullptr;
	}

	for (const auto& x : Utils->LinkBlocks)
	{
		if (!InspIRCd::Match(sname, x->Name))
			continue;

		if (!ComparePass(*x, password))
		{
			ServerInstance->SNO.WriteToSnoMask('l', "Invalid password on link: {}", x->Name);
			continue;
		}

		if (!CheckDuplicate(sname, sid))
			return nullptr;

		const SSLIOHook* const ssliohook = SSLIOHook::IsSSL(this);
		if (ssliohook)
		{
			std::string ciphersuite;
			ssliohook->GetCiphersuite(ciphersuite);
			ServerInstance->SNO.WriteToSnoMask('l', "Negotiated ciphersuite {} on link {}", ciphersuite, x->Name);
		}
		else if (!IsLocalRange(capab->remotesa))
		{
			this->SendError("Non-local server connections MUST be linked with SSL!");
			return nullptr;
		}

		ServerInstance->SNO.WriteToSnoMask('l', "Verified server connection " + linkID + " ("+description+")");
		return x;
	}

	this->SendError("Mismatched server name or password (check the other server's snomask output for details - e.g. user mode +s +Ll)");
	ServerInstance->SNO.WriteToSnoMask('l', "Server connection from \002"+sname+"\002 denied, invalid link credentials");
	return nullptr;
}

/*
 * This is used after the other side of a connection has accepted our credentials.
 * They are then introducing themselves to us, BEFORE either of us burst. -- w
 */
bool TreeSocket::Outbound_Reply_Server(CommandBase::Params& params)
{
	const std::shared_ptr<Link> x = AuthRemote(params);
	if (x)
	{
		/*
		 * They're in WAIT_AUTH_2 (having accepted our credentials).
		 * Set our state to CONNECTED (since everything's peachy so far) and send our
		 * netburst to them, which will trigger their CONNECTED state, and BURST in reply.
		 *
		 * While we're at it, create a treeserver object so we know about them.
		 *   -- w
		 */
		FinishAuth(params[0], params[proto_version == PROTO_INSPIRCD_3 ? 3 : 2], params.back(), x->Hidden);

		return true;
	}

	return false;
}

bool TreeSocket::CheckDuplicate(const std::string& sname, const std::string& sid)
{
	// Check if the server name is not in use by a server that's already fully connected
	TreeServer* CheckDupe = Utils->FindServer(sname);
	if (CheckDupe)
	{
		std::string pname = CheckDupe->GetParent() ? CheckDupe->GetParent()->GetName() : "<ourself>";
		SendError("Server "+sname+" already exists on server "+pname+"!");
		ServerInstance->SNO.WriteToSnoMask('l', "Server connection from \002"+sname+"\002 denied, already exists on server "+pname);
		return false;
	}

	// Check if the id is not in use by a server that's already fully connected
	ServerInstance->Logs.Debug(MODNAME, "Looking for dupe SID {}", sid);
	CheckDupe = Utils->FindServerID(sid);

	if (CheckDupe)
	{
		this->SendError("Server ID "+CheckDupe->GetId()+" already exists on server "+CheckDupe->GetName()+"! You may want to specify the server ID for the server manually with <server:id> so they do not conflict.");
		ServerInstance->SNO.WriteToSnoMask('l', "Server connection from \002"+sname+"\002 denied, server ID '"+CheckDupe->GetId()+
				"' already exists on server "+CheckDupe->GetName());
		return false;
	}

	return true;
}

/*
 * Someone else is attempting to connect to us if this is called. Validate their credentials etc.
 *		-- w
 */
bool TreeSocket::Inbound_Server(CommandBase::Params& params)
{
	const std::shared_ptr<Link> x = AuthRemote(params);
	if (x)
	{
		// Save these for later, so when they accept our credentials (indicated by BURST) we remember them
		this->capab->hidden = x->Hidden;
		this->capab->sid = params[proto_version == PROTO_INSPIRCD_3 ? 3 : 2];
		this->capab->description = params.back();
		this->capab->name = params[0];

		// Send our details: Our server name and description and hopcount of 0,
		// along with the sendpass from this block.
		this->WriteLine(INSP_FORMAT("SERVER {} {} {}{} :{}",
			ServerInstance->Config->ServerName,
			TreeSocket::MakePass(x->SendPass, this->GetTheirChallenge()),
			proto_version == PROTO_INSPIRCD_3 ? "0 " : "",
			ServerInstance->Config->ServerId,
			ServerInstance->Config->ServerDesc
		));

		// move to the next state, we are now waiting for THEM.
		this->LinkState = WAIT_AUTH_2;
		return true;
	}

	return false;
}

CommandServer::Builder::Builder(TreeServer* server)
	: CmdBuilder(server->GetParent(), "SERVER")
{
	push(server->GetName());
	push(server->GetId());
	if (server->IsBursting())
		push_property("burst", ConvToStr(server->StartBurst));
	push_property("hidden", ConvToStr(server->Hidden));
	push_last(server->GetDesc());
}
